libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1027 @@
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected entry node to be an object, but was %s", node_str);
+
+      return NULL;
+    }
+
+  action_type_str = json_object_get_string_member (object, "action");
+
+  if (action_type_str == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected an 'action' member in the autoinstall spec (at %s)", node_str);
+
+      return NULL;
+    }
+
+  if (!flatpak_remote_ref_action_type_parse (action_type_str, &action_type, error))
+    return NULL;
+
+  /* XXX: How do we validate that this value is sensible, as far as I can
+   * tell, json_object_get_int_member would return 0 if the member was not
+   * set or was not an int */
+  serial_node = json_object_get_member (object, "serial");
+  if (serial_node == NULL ||
+      !JSON_NODE_HOLDS_VALUE (serial_node) ||
+      json_node_get_value_type (serial_node) != G_TYPE_INT64)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected an 'index' member in the autoinstall spec (at %s)", node_str);
+      return NULL;
+    }
+
+  serial = json_node_get_int (serial_node);
+
+  detail_object = json_object_get_object_member (object, "detail");
+
+  if (detail_object == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected a 'detail' member in the autoinstall spec (at %s)", node_str);
+      return NULL;
+    }
+
+  flatpak_remote_ref = flatpak_remote_ref_from_action_detail (action_type, detail_object, &local_error);
+
+  if (!flatpak_remote_ref)
+    {
+      if (g_error_matches (local_error,
+                           EOS_UPDATER_ERROR,
+                           EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC))
+        {
+          g_autofree gchar *node_str = json_node_to_string (node);
+          g_propagate_prefixed_error (error,
+                                      local_error,
+                                      "Error parsing the 'detail' member for action (at %s) '%s': ",
+                                      action_type_str,
+                                      node_str);
+          return NULL;
+        }
+
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return NULL;
+    }
+  
+  return flatpak_remote_ref_action_new (action_type, flatpak_remote_ref, serial);
+}
+
+static gint
+sort_flatpak_remote_ref_actions (gconstpointer a, gconstpointer b)
+{
+  FlatpakRemoteRefAction **action_a = (FlatpakRemoteRefAction **) a;
```

Comment:
I don't think you need a cast here

Reply:
It looks like the cast is necessary because the input arguments passed to the function are `const`